File: interop/harness.go
package main

import (
	"encoding/hex"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const version = "1.0.0"

// TestVector represents a single interop test case
type TestVector struct {
	Name         string            `json:"name"`
	Description  string            `json:"description,omitempty"`
	Method       string            `json:"method"` // "rpk" or "pkix"
	TestType     string            `json:"test_type"`
	ECHConfig    string            `json:"ech_config"`
	SigningKey   *SigningKey       `json:"signing_key,omitempty"`
	Signature    *Signature        `json:"signature"`
	Verification *Verification     `json:"verification"`
	Expected     ExpectedResult    `json:"expected"`
	Source       string            `json:"source,omitempty"`
}

type SigningKey struct {
	Algorithm      string `json:"algorithm"`
	PrivateKeyHex  string `json:"private_key_hex"`
	SPKIHex        string `json:"spki_hex"`
}

type Signature struct {
	Algorithm         uint16 `json:"algorithm"`
	NotAfter          uint64 `json:"not_after"`
	AuthenticatorHex  string `json:"authenticator_hex"`
	SignatureHex      string `json:"signature_hex"`
}

type Verification struct {
	CurrentTime  uint64       `json:"current_time"`
	TrustAnchors TrustAnchors `json:"trust_anchors"`
}

type TrustAnchors struct {
	RPKSPKIHashes []string `json:"rpk_spki_hashes,omitempty"`
	PKIXRoots     []string `json:"pkix_roots,omitempty"`
}

type ExpectedResult struct {
	Valid              bool   `json:"valid"`
	ErrorContains      string `json:"error_contains,omitempty"`
	SPKIHashHex        string `json:"spki_hash_hex,omitempty"`
	SignedECHConfigHex string `json:"signed_ech_config_hex,omitempty"`
}

// TestVectorFile represents the root JSON structure
type TestVectorFile struct {
	Version     string       `json:"version"`
	TestVectors []TestVector `json:"test_vectors"`
}

// Implementation represents a testable ECH Auth implementation
type Implementation struct {
	Name       string
	VerifyCmd  []string // Command to verify a signature
	SignCmd    []string // Command to sign (optional)
	WorkingDir string
}

type TestResult struct {
	TestName       string
	Implementation string
	Passed         bool
	Error          string
	Duration       time.Duration
}

func main() {
	var (
		vectorFile = flag.String("vectors", "test-vectors/interop.json", "Path to test vector JSON file")
		implFilter = flag.String("impl", "", "Only test specific implementation (rust/go/nss)")
		testFilter = flag.String("test", "", "Only run tests matching this name substring")
		verbose    = flag.Bool("v", false, "Verbose output")
		listTests  = flag.Bool("list", false, "List available tests and exit")
	)
	flag.Parse()

	// Load test vectors
	data, err := os.ReadFile(*vectorFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading test vectors: %v\n", err)
		os.Exit(1)
	}

	var tvFile TestVectorFile
	if err := json.Unmarshal(data, &tvFile); err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing test vectors: %v\n", err)
		os.Exit(1)
	}

	if *listTests {
		fmt.Printf("Test vectors from %s (version %s):\n", *vectorFile, tvFile.Version)
		for i, tv := range tvFile.TestVectors {
			fmt.Printf("  %d. %s (%s, %s)\n", i+1, tv.Name, tv.Method, tv.TestType)
			if tv.Description != "" {
				fmt.Printf("     %s\n", tv.Description)
			}
		}
		return
	}

	// Discover implementations
	impls := discoverImplementations()
	if *implFilter != "" {
		filtered := []Implementation{}
		for _, impl := range impls {
			if strings.Contains(strings.ToLower(impl.Name), strings.ToLower(*implFilter)) {
				filtered = append(filtered, impl)
			}
		}
		impls = filtered
	}

	if len(impls) == 0 {
		fmt.Fprintf(os.Stderr, "No implementations found\n")
		os.Exit(1)
	}

	fmt.Printf("ECH Auth Interop Test Harness v%s\n", version)
	fmt.Printf("Test vectors: %s (version %s)\n", *vectorFile, tvFile.Version)
	fmt.Printf("Implementations: %d\n", len(impls))
	fmt.Printf("Test vectors: %d\n\n", len(tvFile.TestVectors))

	// Run tests
	var results []TestResult
	for _, impl := range impls {
		fmt.Printf("Testing %s:\n", impl.Name)
		for _, tv := range tvFile.TestVectors {
			if *testFilter != "" && !strings.Contains(tv.Name, *testFilter) {
				continue
			}

			result := runTest(impl, tv, *verbose)
			results = append(results, result)

			status := "✓ PASS"
			if !result.Passed {
				status = "✗ FAIL"
			}
			fmt.Printf("  %s %s (%dms)\n", status, tv.Name, result.Duration.Milliseconds())
			if !result.Passed && result.Error != "" {
				fmt.Printf("      Error: %s\n", result.Error)
			}
		}
		fmt.Println()
	}

	// Summary
	passed := 0
	failed := 0
	for _, r := range results {
		if r.Passed {
			passed++
		} else {
			failed++
		}
	}

	fmt.Printf("=== Summary ===\n")
	fmt.Printf("Total: %d tests\n", len(results))
	fmt.Printf("Passed: %d\n", passed)
	fmt.Printf("Failed: %d\n", failed)

	if failed > 0 {
		os.Exit(1)
	}
}

func discoverImplementations() []Implementation {
	var impls []Implementation

	// Check for Go implementation
	if _, err := os.Stat("go/go.mod"); err == nil {
		impls = append(impls, Implementation{
			Name:       "Go",
			VerifyCmd:  []string{"go", "run", "./cmd/echauth-verify"},
			SignCmd:    []string{"go", "run", "./cmd/echauth-sign"},
			WorkingDir: "go",
		})
	}

	// Check for Rust implementation
	if _, err := os.Stat("rust/Cargo.toml"); err == nil {
		impls = append(impls, Implementation{
			Name:       "Rust",
			VerifyCmd:  []string{"cargo", "run", "--quiet", "--bin", "ech-verify", "--"},
			SignCmd:    []string{"cargo", "run", "--quiet", "--bin", "ech-sign", "--"},
			WorkingDir: "rust",
		})
	}

	// Check for NSS implementation
	if _, err := os.Stat("nss/echauth_client"); err == nil {
		impls = append(impls, Implementation{
			Name:       "NSS",
			VerifyCmd:  []string{"./echauth_client"},
			WorkingDir: "nss",
		})
	}

	return impls
}

func runTest(impl Implementation, tv TestVector, verbose bool) TestResult {
	start := time.Now()
	result := TestResult{
		TestName:       tv.Name,
		Implementation: impl.Name,
		Passed:         false,
	}

	// Create temporary test file
	tmpDir, err := os.MkdirTemp("", "echauth-test-*")
	if err != nil {
		result.Error = fmt.Sprintf("failed to create temp dir: %v", err)
		result.Duration = time.Since(start)
		return result
	}
	defer os.RemoveAll(tmpDir)

	// Write signed ECHConfig to file
	signedConfig, err := buildSignedConfig(tv)
	if err != nil {
		result.Error = fmt.Sprintf("failed to build signed config: %v", err)
		result.Duration = time.Since(start)
		return result
	}

	configPath := filepath.Join(tmpDir, "test.ech")
	if err := os.WriteFile(configPath, signedConfig, 0644); err != nil {
		result.Error = fmt.Sprintf("failed to write config: %v", err)
		result.Duration = time.Since(start)
		return result
	}

	// Build verification command (implementation-specific)
	cmd := buildVerifyCommand(impl, tv, configPath)
	if cmd == nil {
		result.Error = "implementation does not support verification"
		result.Duration = time.Since(start)
		return result
	}

	// Run the test
	output, err := cmd.CombinedOutput()
	if verbose {
		fmt.Printf("    Command: %v\n", cmd.Args)
		fmt.Printf("    Output: %s\n", string(output))
	}

	// Check result
	if tv.Expected.Valid {
		if err != nil {
			result.Error = fmt.Sprintf("expected success but got error: %v\nOutput: %s", err, string(output))
		} else {
			result.Passed = true
		}
	} else {
		if err == nil {
			result.Error = "expected failure but verification succeeded"
		} else {
			// Check error message if specified
			if tv.Expected.ErrorContains != "" {
				if strings.Contains(string(output), tv.Expected.ErrorContains) || strings.Contains(err.Error(), tv.Expected.ErrorContains) {
					result.Passed = true
				} else {
					result.Error = fmt.Sprintf("error doesn't contain expected string '%s': %v\nOutput: %s", tv.Expected.ErrorContains, err, string(output))
				}
			} else {
				result.Passed = true
			}
		}
	}

	result.Duration = time.Since(start)
	return result
}

func buildSignedConfig(tv TestVector) ([]byte, error) {
	// If expected signed config is provided, use it
	if tv.Expected.SignedECHConfigHex != "" {
		return hex.DecodeString(tv.Expected.SignedECHConfigHex)
	}

	// Otherwise, we'd need to build it from components
	// For now, this is a placeholder - implementations should provide pre-signed configs
	return nil, fmt.Errorf("signed config not provided in test vector")
}

func buildVerifyCommand(impl Implementation, tv TestVector, configPath string) *exec.Cmd {
	// This is a simplified version - actual implementations will have different CLI interfaces
	// For now, assume a common interface:
	// <cmd> verify --config <file> --time <unix_timestamp> --trust-anchor <spki_hash|cert>

	args := append([]string{}, impl.VerifyCmd...)
	args = append(args, "verify", "--config", configPath)

	if tv.Verification != nil {
		args = append(args, "--time", fmt.Sprintf("%d", tv.Verification.CurrentTime))

		if tv.Method == "rpk" && len(tv.Verification.TrustAnchors.RPKSPKIHashes) > 0 {
			for _, hash := range tv.Verification.TrustAnchors.RPKSPKIHashes {
				args = append(args, "--trust-anchor", hash)
			}
		}
	}

	cmd := exec.Command(args[0], args[1:]...)
	if impl.WorkingDir != "" {
		cmd.Dir = impl.WorkingDir
	}

	return cmd
}

File: interop/harness.py
#!/usr/bin/env python3
"""
ECH Auth Interop Test Harness

Simple test runner that executes test vectors against multiple implementations.
"""

import json
import os
import subprocess
import sys
import tempfile
import time
from pathlib import Path
from typing import Dict, List, Optional, Tuple

VERSION = "1.0.0"


class TestResult:
    def __init__(self, test_name: str, impl_name: str, passed: bool, error: str = "", duration_ms: int = 0):
        self.test_name = test_name
        self.impl_name = impl_name
        self.passed = passed
        self.error = error
        self.duration_ms = duration_ms


class Implementation:
    def __init__(self, name: str, verify_cmd: List[str], working_dir: str = "."):
        self.name = name
        self.verify_cmd = verify_cmd
        self.working_dir = working_dir

    def can_verify(self) -> bool:
        """Check if this implementation supports verification"""
        return len(self.verify_cmd) > 0


def discover_implementations() -> List[Implementation]:
    """Discover available ECH Auth implementations"""
    impls = []

    # Go implementation
    if Path("go/go.mod").exists():
        impls.append(Implementation(
            name="Go",
            verify_cmd=["go", "run", "./cmd/echauth-verify"],
            working_dir="go"
        ))

    # Rust implementation
    if Path("rust/Cargo.toml").exists():
        impls.append(Implementation(
            name="Rust",
            verify_cmd=["cargo", "run", "--quiet", "--bin", "ech-verify", "--"],
            working_dir="rust"
        ))

    # NSS implementation
    if Path("nss/echauth_client").exists():
        impls.append(Implementation(
            name="NSS",
            verify_cmd=["./echauth_client"],
            working_dir="nss"
        ))

    return impls


def load_test_vectors(path: str) -> Tuple[str, List[Dict]]:
    """Load test vectors from JSON file"""
    with open(path, 'r') as f:
        data = json.load(f)
    return data.get('version', 'unknown'), data.get('test_vectors', [])


def build_signed_config(test_vector: Dict) -> Optional[bytes]:
    """Build signed ECHConfig from test vector"""
    # If pre-built signed config is provided, use it
    if 'expected' in test_vector and 'signed_ech_config_hex' in test_vector['expected']:
        hex_str = test_vector['expected']['signed_ech_config_hex']
        return bytes.fromhex(hex_str)
    
    # Otherwise would need to construct from components
    return None


def run_test(impl: Implementation, test_vector: Dict, verbose: bool = False) -> TestResult:
    """Run a single test case"""
    start_time = time.time()
    test_name = test_vector['name']
    
    if not impl.can_verify():
        return TestResult(test_name, impl.name, False, "Implementation does not support verification")
    
    # Create temporary test file
    with tempfile.TemporaryDirectory() as tmpdir:
        config_path = os.path.join(tmpdir, "test.ech")
        
        # Build signed config
        signed_config = build_signed_config(test_vector)
        if signed_config is None:
            return TestResult(test_name, impl.name, False, "Signed config not provided in test vector")
        
        with open(config_path, 'wb') as f:
            f.write(signed_config)
        
        # Build command
        cmd = list(impl.verify_cmd)
        cmd.extend(["verify", "--config", config_path])
        
        # Add verification parameters
        if 'verification' in test_vector:
            verification = test_vector['verification']
            if 'current_time' in verification:
                cmd.extend(["--time", str(verification['current_time'])])
            
            if test_vector['method'] == 'rpk':
                trust_anchors = verification.get('trust_anchors', {})
                for spki_hash in trust_anchors.get('rpk_spki_hashes', []):
                    cmd.extend(["--trust-anchor", spki_hash])
        
        # Run command
        try:
            result = subprocess.run(
                cmd,
                cwd=impl.working_dir,
                capture_output=True,
                text=True,
                timeout=30
            )
            
            if verbose:
                print(f"    Command: {' '.join(cmd)}")
                print(f"    Output: {result.stdout}")
                if result.stderr:
                    print(f"    Stderr: {result.stderr}")
            
            # Check result
            expected = test_vector.get('expected', {})
            should_pass = expected.get('valid', False)
            
            if should_pass:
                if result.returncode == 0:
                    passed = True
                    error = ""
                else:
                    passed = False
                    error = f"Expected success but got error: {result.stderr}"
            else:
                if result.returncode != 0:
                    # Check error message if specified
                    error_contains = expected.get('error_contains', '')
                    if error_contains:
                        if error_contains in result.stdout or error_contains in result.stderr:
                            passed = True
                            error = ""
                        else:
                            passed = False
                            error = f"Error doesn't contain expected string '{error_contains}': {result.stderr}"
                    else:
                        passed = True
                        error = ""
                else:
                    passed = False
                    error = "Expected failure but verification succeeded"
            
            duration_ms = int((time.time() - start_time) * 1000)
            return TestResult(test_name, impl.name, passed, error, duration_ms)
            
        except subprocess.TimeoutExpired:
            return TestResult(test_name, impl.name, False, "Test timed out after 30 seconds")
        except Exception as e:
            return TestResult(test_name, impl.name, False, f"Exception: {str(e)}")


def main():
    import argparse
    parser = argparse.ArgumentParser(description='ECH Auth Interop Test Harness')
    parser.add_argument('--vectors', default='test-vectors/interop.json',
                       help='Path to test vector JSON file')
    parser.add_argument('--impl', help='Only test specific implementation')
    parser.add_argument('--test', help='Only run tests matching this name')
    parser.add_argument('-v', '--verbose', action='store_true',
                       help='Verbose output')
    parser.add_argument('--list', action='store_true',
                       help='List available tests and exit')
    
    args = parser.parse_args()
    
    # Load test vectors
    try:
        tv_version, test_vectors = load_test_vectors(args.vectors)
    except Exception as e:
        print(f"Error loading test vectors: {e}", file=sys.stderr)
        return 1
    
    if args.list:
        print(f"Test vectors from {args.vectors} (version {tv_version}):")
        for i, tv in enumerate(test_vectors, 1):
            method = tv.get('method', 'unknown')
            test_type = tv.get('test_type', 'unknown')
            print(f"  {i}. {tv['name']} ({method}, {test_type})")
            if 'description' in tv:
                print(f"     {tv['description']}")
        return 0
    
    # Discover implementations
    impls = discover_implementations()
    if args.impl:
        impls = [impl for impl in impls if args.impl.lower() in impl.name.lower()]
    
    if not impls:
        print("No implementations found", file=sys.stderr)
        return 1
    
    print(f"ECH Auth Interop Test Harness v{VERSION}")
    print(f"Test vectors: {args.vectors} (version {tv_version})")
    print(f"Implementations: {len(impls)}")
    print(f"Test vectors: {len(test_vectors)}\n")
    
    # Run tests
    all_results = []
    for impl in impls:
        print(f"Testing {impl.name}:")
        for tv in test_vectors:
            if args.test and args.test not in tv['name']:
                continue
            
            result = run_test(impl, tv, args.verbose)
            all_results.append(result)
            
            status = "✓ PASS" if result.passed else "✗ FAIL"
            print(f"  {status} {tv['name']} ({result.duration_ms}ms)")
            if not result.passed and result.error:
                print(f"      Error: {result.error}")
        print()
    
    # Summary
    passed = sum(1 for r in all_results if r.passed)
    failed = sum(1 for r in all_results if not r.passed)
    
    print("=== Summary ===")
    print(f"Total: {len(all_results)} tests")
    print(f"Passed: {passed}")
    print(f"Failed: {failed}")
    
    return 1 if failed > 0 else 0


if __name__ == '__main__':
    sys.exit(main())
